### ISSUE TITLE
Black level correction

### ISSUE THREAD

**Issue by rwb27** [6by9/lens_shading#2]
In my various messings-around with lens shading, I noticed some strange things happening with the lens shading table.  I think at least some of these strange effects are due to the black level in the raw image: the IMX219 (don't know about the older chip) has a black level of 64 in the raw image, I guess so that the built-in black level compensation doesn't produce negative values/overflow.  Subtracting 64 before dividing seems to fix it, though I've only tested this in my Python/Numpy code, not in your C example.  I suspect you could fix it, however, by changing line 308 to read:

```
				int gain = (middle_val*3) / (avg - 64*3);
```
Thanks for this example code - without it I would not even have gotten started.

**Comment by 6by9** [comment]
Doh! Thank you, you're totally correct.
The len shading block is post black level correction, therefore the raw image needs to be compensated for in the same way when computing the gains. The range from the black level threshold to max Bayer value (eg 1024 for raw10) is stretched out to 0 to max.
I'd need to confirm that the numbers fall out as expected, but it's probably better to do the same black level correction on the raw data as it is unpacked, rather than munging the computed averages.

imx219 tuning has 66 as the black level, whilst ov5647 is 16. Almost all my testing had been with ov5647, therefore the lower black level would explain why I didn't notice any significant errors.

**Comment by rwb27** [comment]
yes - I only spotted it because, when calibrating with a dimmer-than-ideal scene, the correction was only partial.  It took me an embarrassingly long time to figure out it was a black level thing!

Is there a reason you use 66 rather than 64?  I'm just curious now; I was fairly sure the datasheet said 64, so is the slightly higher number getting around some of the noise?

**Comment by 6by9** [comment]
I'm copying from the tuning data that is programmed into the isp for the relevant sensor, slightly simplified as it uses 66 on some channels and 67 on others.
I don't know how they determined those values during tuning - probably by measuring a sample set of golden sensors.

**Comment by 6by9** [comment]
Could you confirm that the updated tool gives the same or very similar tables to those you had with your change? Based on how the pipe is configured I believe my patch is correct, but only have a couple of raw images to hand to test with.

**Comment by rwb27** [comment]
Thanks - I'm embarrassed to say I've not actually compiled the raspistill binaries myself (which I think is how you get your lens shading header in?), but I have a couple of excellent students who have now done that, and I'll see if I can get them to test it out in the next few weeks.
